renormalizer/mps/mpo.py

```diff
@@ -123,25 +116,39 @@ def symbolic_mpo(table, factor, algo="Hopcroft-Karp"):
     logger.debug(f"Input operator terms: {len(table)}")
     # translate the symbolic operator table to an easy to manipulate numpy array
     # extract the op symbol, qn, factor to a numpy array
-    symbol_table = np.array([[x.symbol for x in ta] for ta in table])
+
+    # np.array can't handle tuple as array element. So use dict instead
+    def op_to_dict(op: Op):
+        return {0: op.symbol, 1: tuple(op.dofs)}
+    symbol_table = np.array([[op_to_dict(x) for x in ta] for ta in table])
     qn_table = np.array([[x.qn for x in ta] for ta in table])
     factor_table = np.array([[x.factor for x in ta] for ta in table])
-    
+
     new_table = np.zeros((len(table), nsite),dtype=np.uint16)
-    
-    primary_ops = {}
-    # check that op with the same symbol has the same factor and qn
-    unique_symbol = np.unique(symbol_table)
+
+    # Construct mapping from easy-to-manipulate integer to actual Op
+    # There has to be an Identity for the algorithm to work
+    # even if Identity is not in the table
+    primary_ops = {0: Op.identity()}
+    # unique operators with DoF names taken into consideration
+    # The inclusion of DoF names is necessary fo multi-dof basis.
+    unique_op: Set[Op] = set(np.array(table).ravel())
+    if primary_ops[0] in unique_op:
+        unique_op.remove(primary_ops[0])
```

Comment:
I don't understand why `Op.identity` is necessary here (restricting the key to be 0).
Though the left-most `in_ops` is an identity operator physically, it is technically just a placeholder with `qn=0` and corresponds to an additional column `np.zeros` attached to `op_table` to make the logic more consistent when looping from 0 to N-1.  
For example primary_ops = {0:xxx, 1: op.indentity()} will also obtain the correct answer.

Reply:
The key could be anything. The problem is that if the operator "identity" is not present in the model, there will be actually no `primary_ops` key for the operator "identity". This is not a serious issue because the algorithm still works as you've pointed out, but I feel somewhat unhappy with the situation. I agree that the resulting code is ugly and hard to grasp. I'll clean it up and add a few comments on the code.

Reply:
ok.